### PR TITLE
toolrunner should which tool

### PR DESCRIPTION
**Pull request by ericsciple**
fixes https://github.com/actions/toolkit/issues/218

### CODE REVIEW

File: packages/exec/src/toolrunner.ts [hunk 1]
```diff
@@ -392,6 +393,8 @@ export class ToolRunner extends events.EventEmitter {
    * @returns   number
    */
   async exec(): Promise<number> {
+    this.toolPath = await io.which(this.toolPath, true)
```
Comment (joshmgross) [inline packages/exec/src/toolrunner.ts]: Is this a breaking change?

Comment (ericsciple) [reply packages/exec/src/toolrunner.ts]: my intention was that it doesnt break anything

for example `which` returns the path if it is already rooted and executable.

however, now that you got me thinking about it more, i think there are some edge cases wrt relative pathing.... i'll add code to handle.

File: packages/exec/README.md [hunk 2]
```diff
@@ -48,13 +48,11 @@ await exec.exec('node', ['index.js', 'foo=bar'], options);
...
+You can specify the full path for tools not in the PATH:
 
 ```js
 const exec = require('@actions/exec');
 const io = require('@actions/io');
 
-const pythonPath: string = await io.which('python', true)
-
-await exec.exec(`"${pythonPath}"`, ['main.py']);
+await exec.exec('"<FULL-PATH-TO-TOOL>"', ['arg1']);
```
Comment (ericsciple) [inline packages/exec/README.md]: update with an example

File: packages/exec/src/toolrunner.ts (hunk 1 above)
Comment (ericsciple) [reply packages/exec/src/toolrunner.ts]: ok i added the code above to root unrooted paths that contain relative pathing

in addition to checking the PATH, the which function also handles checking rooted paths

now the changes shouldnt be a breaking change. hard to imagine it would have practically broken anyone, but i'd rather cover edge cases to be on the safe side. Also now the changes align better with the default behavior of `spawn`.

Comment (ericsciple) [reply packages/exec/src/toolrunner.ts]: i'll add unit tests tomorrow for the unrooted-relative-pathing cases

File: .github/workflows/unit-tests.yml [hunk 3]
```diff
@@ -6,18 +6,27 @@ on:
...
+
+    runs-on: ${{ matrix.runs-on }}
+
     steps:
     - name: Checkout
-      uses: actions/checkout@master
+      uses: actions/checkout@v1
 
-    - name: Set Node.js 10.x
+    - name: Set Node.js 12.x
```
Comment (ericsciple) [inline .github/workflows/unit-tests.yml]: the runner carries node 12

Thankfully this also happens to fix the ENOMEM toolrunner test issues on Windows

File: .github/workflows/unit-tests.yml [hunk 4]
```diff
@@ -6,18 +6,27 @@ on:
...
 jobs:
-  Ubuntu:
-    name: Run Ubuntu
-    runs-on: ubuntu-latest
+
+  build:
+    name: Build
+
+    strategy:
+      matrix:
```
Comment (ericsciple) [inline .github/workflows/unit-tests.yml]: leverage matrix instead of duplicate 3 times

File: package.json [hunk 5]
```diff
@@ -9,8 +9,7 @@
     "format-check": "prettier --check packages/**/*.ts",
     "lint": "eslint packages/**/*.ts",
     "new-package": "scripts/create-package",
-    "test": "jest",
-    "test-ci": "jest --testPathIgnorePatterns=\"<rootDir>/packages/exec/__tests__/exec.test.ts\""
```
Comment (ericsciple) [inline package.json]: we dont need this command anymore, this was to skip the failing tests on windows

File: packages/exec/README.md [hunk 6]
```diff
@@ -48,13 +48,10 @@ await exec.exec('node', ['index.js', 'foo=bar'], options);
...
+You can specify the full path for tools not in the PATH:
 
 ```js
 const exec = require('@actions/exec');
-const io = require('@actions/io');
 
-const pythonPath: string = await io.which('python', true)
-
-await exec.exec(`"${pythonPath}"`, ['main.py']);
+await exec.exec('"/path/to/my-tool"', ['arg1']);
```
Comment (ericsciple) [inline packages/exec/README.md]: the previous example did the exact opposite of the description :(

File: packages/exec/__tests__/exec.test.ts [hunk 7]
```diff
@@ -121,6 +121,38 @@ describe('@actions/exec', () => {
     }
   })
 
+  it('Runs exec successfully with command from PATH', async () => {
```
Comment (ericsciple) [inline packages/exec/__tests__/exec.test.ts]: coverage for existing functionality

File: packages/exec/__tests__/exec.test.ts [hunk 8]
```diff
@@ -418,6 +450,134 @@ describe('@actions/exec', () => {
     fs.unlinkSync(semaphorePath)
   })
 
+  it('Exec roots relative tool path using unrooted options.cwd', async () => {
```
Comment (ericsciple) [inline packages/exec/__tests__/exec.test.ts]: tests to make sure we're not breaking existing behavior

also this confirms that extensions are resolved on windows, which the previous code did not do (but should have done)

File: packages/exec/__tests__/exec.test.ts [hunk 9]
```diff
@@ -572,6 +732,42 @@ describe('@actions/exec', () => {
       )
     })
 
+    it('execs .cmd from path (Windows)', async () => {
```
Comment (ericsciple) [inline packages/exec/__tests__/exec.test.ts]: here is the test related to the bug fix.

toolrunner should which the tool before attempting to invoke. otherwise simple commands like `npm --version` fail on windows (ENOENT) because `npm` resolves to `npm.cmd` and node doesnt find it.

File: packages/exec/__tests__/exec.test.ts [hunk 10]
```diff
@@ -418,6 +450,134 @@ describe('@actions/exec', () => {
     fs.unlinkSync(semaphorePath)
   })
 
+  it('Exec roots relative tool path using unrooted options.cwd', async () => {
+    let exitCode: number
+    let command: string
+    if (IS_WINDOWS) {
+      command = './print-args-cmd' // let ToolRunner resolve the extension
```
Comment (ericsciple) [inline packages/exec/__tests__/exec.test.ts]: note, without extension, toolrunner should resolve the extension

File: packages/exec/__tests__/scripts/print-args-cmd.cmd [hunk 11]
```diff
@@ -0,0 +1,12 @@
+@echo off
```
Comment (ericsciple) [inline packages/exec/__tests__/scripts/print-args-cmd.cmd]: the naming of these script files is a little bit weird, but it's related to the conventions set forth by the existing test case scripts

Comment (ericsciple) [reply packages/exec/__tests__/scripts/print-args-cmd.cmd]: i just copied one of the existing cmd files (but it had spaces in the path, and i didnt want that)

this script basically just echoes the args, e.g.

```
arg[0]: "hello"
arg[1]: "world"
```

File: packages/exec/__tests__/scripts/print-args-sh.sh [hunk 12]
```diff
@@ -0,0 +1,11 @@
+#!/usr/bin/env bash
+
+# store arguments in a special array
+args=("$@")
+# get number of elements
+ELEMENTS=${#args[@]}
+
+# echo each element
+for (( i=0;i<$ELEMENTS;i++)); do
+    echo "args[$i]: \"${args[${i}]}\""
```
Comment (ericsciple) [inline packages/exec/__tests__/scripts/print-args-sh.sh]: same deal here, this script just echos the args

File: packages/exec/package.json [hunk 13]
```diff
@@ -32,7 +32,7 @@
   "bugs": {
     "url": "https://github.com/actions/toolkit/issues"
   },
-  "devDependencies": {
+  "dependencies": {
```
Comment (ericsciple) [inline packages/exec/package.json]: need to call `io.which` and also some other io helpers

File: .github/workflows/unit-tests.yml (hunk 3 above)
Comment (joshmgross) [reply .github/workflows/unit-tests.yml]: Creating actions docs recommends `  using: 'node12'`, do we officially support actions on node 10 and node 12? 
If so, I think it would be worth having the workflow matrix across node versions, perhaps with certain functions explicitly listed as not supported on node 10.

Comment (ericsciple) [reply .github/workflows/unit-tests.yml]: we only use node12 to run actions. I think the runner initially carried node v10, and then switched to node v12 before GA

File: packages/exec/src/toolrunner.ts (hunk 1 above)
Comment (bryanmacfarlane) [reply packages/exec/src/toolrunner.ts]: No, it's not a breaking change.  It's additive.  If you were passing in a fully qualified path, which will not changing anything.  It will however add the ability to just pass in a tool name and it just works (as well as fixing some other issues and bugs).

Comment (joshmgross) [reply packages/exec/src/toolrunner.ts]: @bryanmacfarlane Originally it was breaking, exec of a relative path would no longer work if we automatically `which`. That should be fixed now with @ericsciple's latest change

File: packages/exec/__tests__/exec.test.ts (hunk 8 above)
Comment (thboop) [inline packages/exec/__tests__/exec.test.ts]: .cwd?

File: packages/exec/__tests__/exec.test.ts (hunk 10 above)
Comment (thboop) [inline packages/exec/__tests__/exec.test.ts]: Consider adding a `\\` test for windows

File: packages/exec/src/toolrunner.ts [hunk 14]
```diff
@@ -392,6 +395,24 @@ export class ToolRunner extends events.EventEmitter {
...
    */
   async exec(): Promise<number> {
+    // root the tool path if it is unrooted and contains relative pathing
+    if (
+      !ioUtil.isRooted(this.toolPath) &&
+      (this.toolPath.includes('/') ||
+        (IS_WINDOWS && this.toolPath.includes('\\')))
+    ) {
+      // prefer options.cwd if it is specified, however options.cwd may also need to be rooted
+      this.toolPath = path.resolve(
```
Comment (thboop) [inline packages/exec/src/toolrunner.ts]: this is super clean 🥇

**Comment by thboop** [comment]
We have existing tests for exec.exec ("full path to executable"), these tests should cover the rest of the cases and fix this bug

File: packages/tool-cache/package-lock.json [hunk 15]
```diff
@@ -1,6 +1,6 @@
 {
 	"name": "@actions/tool-cache",
-	"version": "1.1.1",
+	"version": "1.1.2",
```
Comment (1138-4EB) [inline packages/tool-cache/package-lock.json]: The version of `tool-cache` was bumped, even though it seems that none of the files in `packages/tool-cache` was changed. Is this expected?

Should it mean to reflect that `tool-cache` is updated to use this new feature of `exec`, I'd expect files such as `packages/tool-cache/src/tool-cache.ts` to be modified:

https://github.com/actions/toolkit/blob/5c894298f2f24441718eb049215aa6decd507193/packages/tool-cache/src/tool-cache.ts#L175

https://github.com/actions/toolkit/blob/5c894298f2f24441718eb049215aa6decd507193/packages/tool-cache/src/tool-cache.ts#L203

https://github.com/actions/toolkit/blob/5c894298f2f24441718eb049215aa6decd507193/packages/tool-cache/src/tool-cache.ts#L239

https://github.com/actions/toolkit/blob/5c894298f2f24441718eb049215aa6decd507193/packages/tool-cache/src/tool-cache.ts#L254
